src/new_skills/NewSkill.cpp

```diff
@@ -0,0 +1,136 @@
+//
+// Created by jessevw on 21.02.20.
+//
+
+#include "include/roboteam_ai/new_skills/NewSkill.h"
+#include "skills/Skill.h"
+
+#include <include/roboteam_ai/utilities/Settings.h>
+
+#include <cmath>
+
+#include "control/ControlUtils.h"
+#include "utilities/Constants.h"
+#include "utilities/GameStateManager.hpp"
+#include "utilities/RobotDealer.h"
+#include "world/Ball.h"
+#include "world/Robot.h"
+#include "world/World.h"
+
+namespace rtt::ai {
+
+    NewSkill::NewSkill(std::string name, bt::Blackboard::Ptr blackboard) : bt::Leaf(std::move(name), std::move(blackboard)) {
+        robot = std::make_shared<Robot>();
+        ball = std::make_shared<Ball>();
+    }
+
+    void NewSkill::publishRobotCommand() {
+        if (!SETTINGS.isLeft()) {
+            command = rotateRobotCommand(command);
+        }
+
+        limitRobotCommand();
+
+        if (command.id() == -1) {
+            std::cerr << "the command is trying to send a RobotCommand to id " << command.id() << std::endl;
+            if (robot && robot->id != -1) {
+                command.set_id(robot->id);
+                io::io.publishRobotCommand(command);  // We default to our robots being on the left if parameter is not set
+            }
+        } else {
+            io::io.publishRobotCommand(command);  // We default to our robots being on the left if parameter is not set
+        }
+        // refresh the robotcommand after it has been sent
+        refreshRobotCommand();
+    }
+
+    std::string NewSkill::node_name() { return name; }
+
+    NewSkill::Status NewSkill::update() {
+        robotId = -15;// TODO: after the tactic is implemented, needs to be some way the skill figures out which id to send to. Should be known by the tactic
+        updateRobot();
+        ball = world->getBall();  // update ball position
+        if (!robot){
+            std::cerr << "The robot is not initialized" << std::endl;
+            return Status::Failure;
+        }
+        if (robot->id == -1) {
+            std::cerr << "The robot id is" << robot->id << std::endl;
+            return Status::Failure;
+        }
+        if (!ball) {
+            return Status::Waiting;
+        }
+        return onUpdate();
+    }
+
+    void NewSkill::initialize() {
+        // TODO: be able to get robot from dealer
+        // robot = dealer.distributions.at(role).robot;
```

Comment:
I'm really curious how this would work. I think it would be better if the skill didn't even have access to the robot - just its ID. I suppose that the bt::Leaf already has a reference to a robot

Reply:
Yeah it is definitely not meant to happen here, imo this should happen at play strategic level and then each role should be coupled to an ID, after which the roles are executed. Each role has some priority decided beforehand in the form of flags so it can be best assigned to a robot